src/status.rs

```diff
@@ -309,9 +417,320 @@ mod test {
+                    new_state: Some(new_state),
+                    approve_state: None,
+                }),
+                heartbeats: Some(vec![heartbeat_1]),
+            },
+            follower: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: None,
+                    approve_state: Some(approve_state),
+                }),
+                heartbeats: Some(vec![heartbeat_2]),
+            },
+            recency: Duration::minutes(4),
+        };
+
+        assert_eq!(
+            is_disconnected(&channel, &messages),
+            true,
+            "Follower heartbeats have no recent messages"
+        )
+    }
+
+    #[test]
+    fn no_recent_leader_hb_on_follower_validator() {
+        let server = SERVER_POOL.get_server();
+        let channel = get_test_channel(&server);
+        let follower_id = &channel.spec.validators.follower().id;
+
+        let mut heartbeat_1 = get_heartbeat_msgs(Duration::minutes(0));
+        heartbeat_1.from = follower_id.clone();
+        let mut heartbeat_2 = get_heartbeat_msgs(Duration::minutes(0));
+        heartbeat_2.from = follower_id.clone();
+        let new_state = get_newstate_msg();
+        let approve_state = get_approve_state_msg();
+
+        let messages = Messages {
+            leader: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: Some(new_state),
+                    approve_state: None,
+                }),
+                heartbeats: Some(vec![]),
+            },
+            follower: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: None,
+                    approve_state: Some(approve_state),
+                }),
+                heartbeats: Some(vec![heartbeat_1, heartbeat_2]),
+            },
+            recency: Duration::minutes(4),
+        };
+
+        assert_eq!(
+            is_disconnected(&channel, &messages),
+            true,
+            "Follower validator has recent HB messages but none have come from the leader validator"
+        )
+    }
+
+    #[test]
+    fn recent_hbs_on_both_arrays() {
+        let server = SERVER_POOL.get_server();
+        let channel = get_test_channel(&server);
+        let mut heartbeat_1 = get_heartbeat_msgs(Duration::minutes(0));
+        heartbeat_1.from = channel.spec.validators.leader().clone().id;
+        let mut heartbeat_2 = get_heartbeat_msgs(Duration::minutes(0));
+        heartbeat_2.from = channel.spec.validators.follower().clone().id;
+        let new_state = get_newstate_msg();
+        let approve_state = get_approve_state_msg();
+
+        let messages = Messages {
+            leader: LastApprovedResponse {
+                last_approved: Some(LastApproved {
+                    new_state: Some(new_state),
+                    approve_state: None,
+                }),
+                heartbeats: Some(vec![]),
```

Comment:
Missing heartbeat? By the looks of the test there should be 1 heartbeat in the leader and 1 in the follower.